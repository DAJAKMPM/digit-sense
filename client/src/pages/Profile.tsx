import React, {
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState,
} from "react";
import UploadIcon from "@mui/icons-material/Upload";
import { Avatar, Box, Button, Grid, Link, Typography } from "@mui/material";
import { Layout } from "../components/Layout";
import * as API from "../api/Api";
import { User, UserResponse } from "../types";
import { uploadAvatar } from "../services";

const ProfileScreen = () => {
  const [user, setUser] = useState<User | null>(null);
  const inputFileRef = useRef<HTMLInputElement>(null);

  const handleClick = () => inputFileRef.current?.click();

  const handleFetchLoggedInUser = useCallback(async () => {
    try {
      const response = (await API.get("auth/currentUser")) as UserResponse;

      if (response.success) setUser(response.data);
    } catch (error) {
      console.log(error);
    }
  }, []);

  useEffect(() => {
    handleFetchLoggedInUser();
  }, [handleFetchLoggedInUser]);

  const handleOnChange = useCallback(
    async (event: React.ChangeEvent<HTMLInputElement>) => {
      const newImages = event.target?.files;

      if (newImages) {
        try {
          const { success } = await uploadAvatar(newImages);

          if (success) handleFetchLoggedInUser();
        } catch (error) {
          console.log(error);
        }
      }
    },
    [handleFetchLoggedInUser]
  );

  const profilePicture = useMemo(
    () =>
      user?.avatar && user?.avatar !== "no-photo.jpg"
        ? `${process.env.REACT_APP_API_URL}/uploads/${user?.avatar}`
        : `${process.env.REACT_APP_API_URL}/uploads/noimg.jpg`,
    [user]
  );

  return (
    <Layout pageTitle="Profile">
      <Box
        sx={{
          marginTop: 8,
        }}
      >
        <Link underline="hover" color="inherit" href="/">
          Dashboard
        </Link>
      </Box>

      <Box
        sx={{
          marginTop: 8,
        }}
      >
        <Grid container spacing={2}>
          <Grid item xs={12}>
            <Avatar
              alt="Remy Sharp"
              src={profilePicture}
              sx={{ width: 156, height: 156 }}
            />

            <input
              ref={inputFileRef}
              accept="image/*"
              hidden
              id="avatar-image-upload"
              type="file"
              onChange={handleOnChange}
            />
            <label htmlFor="avatar-image-upload">
              <Button
                variant="contained"
                startIcon={<UploadIcon />}
                onClick={handleClick}
                sx={{ mt: 3, mb: 2 }}
              >
                Upload
              </Button>
            </label>
          </Grid>
          <Grid item xs={12}>
            <Typography component="p" variant="h6">
              Email: {user?.email}
            </Typography>
          </Grid>
          <Grid item xs={12}>
            <Typography component="p" variant="h6">
              Full Name: {user?.name}
            </Typography>
          </Grid>
          <Grid item xs={12}>
            <Typography component="p" variant="h6">
              Age: {user?.age}
            </Typography>
          </Grid>
          <Grid item xs={12}>
            <Typography component="p" variant="h6">
              Age: {user?.address}
            </Typography>
          </Grid>
        </Grid>
      </Box>
    </Layout>
  );
};

export { ProfileScreen };
