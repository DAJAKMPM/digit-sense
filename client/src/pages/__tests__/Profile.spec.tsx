import { render } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import { ProfileScreen } from "../Profile";

test("should match the ProfileScreen snapshot properly", () => {
  const { container } = render(
    <Router>
      <ProfileScreen />
    </Router>
  );

  expect(container).toMatchSnapshot();
});
