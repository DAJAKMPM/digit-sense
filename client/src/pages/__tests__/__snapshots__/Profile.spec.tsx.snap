// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the ProfileScreen snapshot properly 1`] = `
<div>
  <main>
    <header
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiAppBar-root MuiAppBar-colorDefault MuiAppBar-positionAbsolute css-3e2rzi-MuiPaper-root-MuiAppBar-root"
    >
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
      >
        <h6
          class="MuiTypography-root MuiTypography-h6 MuiTypography-noWrap css-uh02xs-MuiTypography-root"
        >
          Profile
        </h6>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textInherit MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorInherit css-1y942vo-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Profile
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textInherit MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorInherit css-1y942vo-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Logout
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </header>
    <main
      class="MuiContainer-root MuiContainer-maxWidthLg css-2kezlg-MuiContainer-root"
    >
      <div
        class="MuiBox-root css-h8m1q6"
      >
        <a
          class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineHover css-le90w3-MuiTypography-root-MuiLink-root"
          href="/"
        >
          Dashboard
        </a>
      </div>
      <div
        class="MuiBox-root css-h8m1q6"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
          >
            <div
              class="MuiAvatar-root MuiAvatar-circular css-rf3q7b-MuiAvatar-root"
            >
              <img
                alt="Remy Sharp"
                class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
                src="http://localhost:8000/uploads/noimg.jpg"
              />
            </div>
            <input
              accept="image/*"
              hidden=""
              id="avatar-image-upload"
              type="file"
            />
            <label
              for="avatar-image-upload"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-1n7zltw-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="UploadIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M5 20h14v-2H5v2zm0-10h4v6h6v-6h4l-7-7-7 7z"
                    />
                  </svg>
                </span>
                Upload
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </label>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
          >
            <p
              class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
            >
              Email: 
            </p>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
          >
            <p
              class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
            >
              Full Name: 
            </p>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
          >
            <p
              class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
            >
              Age: 
            </p>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
          >
            <p
              class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
            >
              Age: 
            </p>
          </div>
        </div>
      </div>
    </main>
  </main>
</div>
`;
